keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -215,6 +221,84 @@ class SqlArtifactRepository(
       .execute()
   }
 
+  override fun versionsByEnvironment(deliveryConfig: DeliveryConfig): List<EnvironmentArtifactsSummary> {
+    return deliveryConfig.environments.map { environment ->
```

Comment:
It looks like the response is entirely built from the `ENVIRONMENT_ARTIFACT_VERSIONS` table, so another approach could be to: 

1) get all envIds for the delivery config
2) select * from ENVIRONMENT_ARTIFACT_VERSIONS where ENVIRONMENT_UID in (....) and I think `versionsInEnvironment` and `pendingVersions` can both be built from that. I've found having jooq fetch into a `ResultSet` is helpful when deriving disparate things from a single query. 

I think it would still make sense to have a limit parameter on the method in this case, even though it would have to be implemented in memory, if only to limit the cost to browsers parsing the response.

Reply:
It has to union that with a read from `DELIVERY_ARTIFACT_VERSION` where `NOT EXISTS` in `ENVIRONMENT_ARTIFACT_VERSIONS` because versions only appear on that latter table once they get approved.